crates/topos-p2p/src/behaviour/grpc.rs

```diff
@@ -0,0 +1,516 @@
+                    id,
+                    address,
+                    request_id: Some(request_id),
+                    channel: Some(channel),
+                }) => OutboundConnection::Connected(OutboundConnectedConnection {
+                    request_id: *request_id,
+                    channel: channel.clone(),
+                }),
+                Some(Connection {
+                    id,
+                    address,
+                    request_id: Some(request_id),
+                    channel,
+                }) => {
+                    debug!("Peer already connected but no channel bound");
+
+                    OutboundConnection::Pending {
+                        request_id: *request_id,
+                    }
+                }
+                Some(_) => self.open_connection(peer_id),
+                _ => {
+                    debug!("No connection for this peer {}", peer_id);
+                    self.open_connection(peer_id)
+                }
+            }
+        } else {
+            debug!("Buffering sender as no available connection to peer {peer_id} yet");
+            self.open_connection(peer_id)
+        }
+    }
+
+    /// Return the next outbound request id
+    fn next_request_id(&mut self) -> RequestId {
+        let request_id = self.next_request_id;
+        self.next_request_id.0 += 1;
+
+        request_id
+    }
+
+    /// Try to open a connection with the given peer.
+    fn open_connection(&mut self, peer_id: &PeerId) -> OutboundConnection {
+        info!("Opening gRPC outbound connection to peer {peer_id}");
+
+        let (notifier, receiver) = oneshot::channel();
+        let request_id = self.next_request_id();
+
+        self.pending_outbound_connections
+            .entry(*peer_id)
+            .or_insert_with(|| OutboundConnectionRequest {
+                request_id,
+                notifier,
+            });
+
+        self.pending_events.push_back(ToSwarm::Dial {
+            opts: DialOpts::peer_id(*peer_id).build(),
+        });
+
+        OutboundConnection::Opening {
+            request_id,
+            receiver,
+        }
+    }
+
+    /// Handle the [`ConnectionEstablished`] event comming from the [`Swarm`]
+    /// and try to open a gRPC channel using a [`ConnectionHandler`].
+    fn on_connection_established(
+        &mut self,
+        ConnectionEstablished {
+            peer_id,
+            connection_id,
+            endpoint,
+            failed_addresses,
+            other_established,
+        }: ConnectionEstablished,
+    ) {
+        let address = match endpoint {
+            ConnectedPoint::Dialer { address, .. } => Some(address.clone()),
+            ConnectedPoint::Listener { .. } => None,
+        };
+
+        let connection = Connection {
+            id: connection_id,
+            address,
+            request_id: None,
+            channel: None,
+        };
+
+        self.connected.entry(peer_id).or_default().push(connection);
```

Comment:
Would it work to use `or_insert_with()` here too?

Reply:
`or_insert_with` would need to provide an empty `vec` and then `push` to it?
It is more clear and recommanded to rely on `Default`

Reply:
No, you'd build the content right away so no push needed. `self.connected.entry(peer_id).or_insert_with(|| vec![connection])` works I think.

I don't think it matters, but as a reader, the `or_default()` forced me to go look up what the default is, which isn't relevant in this case.

Reply:
yeah but the behaviour is different and doesn't cover what is expected. Look this playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=10a11a2cea226ce8a965d1e0a58ad9ae